File: packages/sdk-kotlin/src/main/kotlin/org/melaya/ProjectsAPI.kt
package org.melaya

import org.json.JSONObject

/**
 * Projects API — create and list agent projects.
 *
 * Projects are the top-level namespace for pipelines, connectors, and team membership.
 *
 * Paths:
 *   - `GET   /api/v1/private/projects`        — list accessible projects
 *   - `POST  /api/v1/private/projects`        — create a new project
 *   - `PATCH /api/v1/private/projects/rename` — rename a project
 *   - `GET   /api/v1/private/projects/runner` — runner-facing project list
 *
 * @example
 * ```kotlin
 * val projects = melaya.projects.list()
 * val created  = melaya.projects.create(name = "my-agents", description = "Production agents")
 * melaya.projects.rename(oldName = "my-agents", newName = "prod-agents")
 * ```
 */
class ProjectsAPI internal constructor(private val http: HttpClient) {

    /**
     * List all projects the authenticated user can access
     * (owned projects + projects they are a member of).
     */
    fun list(): List<JSONObject> {
        val r = http.get("/api/v1/private/projects")
        return when (r) {
            is org.json.JSONArray -> r.toJsonObjects()
            is JSONObject -> r.optJSONArray("projects")?.toJsonObjects() ?: emptyList()
            else -> emptyList()
        }
    }

    /** Create a new agent project. */
    fun create(name: String, description: String? = null): JSONObject {
        val body = buildMap<String, Any?> {
            put("name", name)
            if (description != null) put("description", description)
        }
        return http.post("/api/v1/private/projects", body).asObject()
    }

    /** Rename a project from [oldName] to [newName]. */
    fun rename(oldName: String, newName: String): JSONObject {
        return http.patch(
            "/api/v1/private/projects/rename",
            mapOf("oldName" to oldName, "newName" to newName)
        ).asObject()
    }

    /** Get the runner-facing project list (used internally by the runner CLI). */
    fun runnerProjects(): List<JSONObject> {
        val r = http.get("/api/v1/private/projects/runner")
        return when (r) {
            is org.json.JSONArray -> r.toJsonObjects()
            is JSONObject -> r.optJSONArray("projects")?.toJsonObjects() ?: emptyList()
            else -> emptyList()
        }
    }
}

File: packages/sdk-csharp/Melaya/ProjectsApi.cs
using System.Text.Json;

namespace Melaya;

/// <summary>
/// Projects API — create and list agent projects.
/// Maps to <c>/api/v1/private/projects</c>.
/// </summary>
public sealed class ProjectsApi
{
    private readonly MelayaHttpClient _http;

    internal ProjectsApi(MelayaHttpClient http) => _http = http;

    /// <summary>List all projects the authenticated user can access (owned + team member).</summary>
    public async Task<List<Project>> ListAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<Project>>("/api/v1/private/projects", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Create a new agent project.</summary>
    public async Task<Project> CreateAsync(ProjectCreateRequest body, CancellationToken ct = default)
    {
        return await _http.PostAsync<Project>("/api/v1/private/projects", body, ct).ConfigureAwait(false);
    }

    /// <summary>Rename a project.</summary>
    public async Task<BoolResult> RenameAsync(string oldName, string newName, CancellationToken ct = default)
    {
        var body = new { oldName, newName };
        return await _http.PatchAsync<BoolResult>("/api/v1/private/projects/rename", body, ct).ConfigureAwait(false);
    }

    /// <summary>Get projects list (runner-facing endpoint).</summary>
    public async Task<JsonElement> RunnerProjectsAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<JsonElement>("/api/v1/private/projects/runner", ct: ct).ConfigureAwait(false);
    }
}
